### DISCUSSION TITLE
Rule Matching NER: Match and label entities only with specific context

### DISCUSSION THREAD

**Discussion by camipower** [explosion/spaCy#8998] [Help: Coding & Implementations] [answered]
Hi!

is it possible to match and label an entity only with specific context? (e.g. match and label "NOUN NOUN", only if there is a DET or ADP before and an ADP or a "dot" after it). 

My actual solution was to write a function to edit the Spans and add it to the pipeline after entity_ruler (using @Language.component). It works, but if I have a sentence like: "DET NOUN NOUN ADP NOUN NOUN ." only the first "NOUN NOUN" would be matched (actually "DET NOUN NOUN ADP", but my function edits the Spans removing both DET and ADP) , but the second "NOUN NOUN" not, because the ADP is already being matched by the first one and being edited by my function after entity_ruler.

One more question:
I would also like to know if it is possible to add a PhraseMatcher to the nlp pipeline and also label the matches for NER? I have a long list of terms I would like to match (case insensitive) and label in the same pipeline together with the entity_ruler I wrote about before. 
I know I could add each phrase as a pattern to the entity_ruler like this:

    {"label": "MY_ENTITY", "pattern": "one important entity"}

but I don't know how to match case-insensitive, without having to do something like this for each item in my list:

    {"label": "MY_ENTITY", "pattern": [{"LOWER": "one"}, {"LOWER": "important"}, {"LOWER": "entity"}]}

I am adding these parts to the nlp pipeline like this:

```
patterns = [pattern1, pattern2, ...]
ruler = nlp.add_pipe("entity_ruler")
ruler.add_patterns(patterns)
nlp.add_pipe("remove_ctx_from_term", after="entity_ruler")  # here comes the function to edit the Spans
doc = nlp(data) ....
```

Thank you for your help/ideas!

**Comment by polm** [comment, marked as answer]
> is it possible to match and label an entity only with specific context? (e.g. match and label "NOUN NOUN", only if there is a DET or ADP before and an ADP or a "dot" after it).

The best way to do this right now is to write a broad matcher and then check the spans that match and use code to discard spans that don't meet your conditions. There is a longstanding wishlist item about having non-capturing matches (#6420) but we're not actively working on it at the moment. You might also be able to use the recently added [match alignments](https://spacy.io/api/matcher#:~:text=return%20match%20alignment%20information) feature.

> I would also like to know if it is possible to add a PhraseMatcher to the nlp pipeline and also label the matches for NER?

You can't add a PhraseMatcher to a pipeline directly, I guess you mean EntityRuler? You can have both an EntityRuler and NER component (or multiples of either) in the same pipeline. Ordering affects how they interact; normally later components won't overwrite earlier components, though you can specifically enable that for the EntityRuler. See the [tutorial project](https://github.com/explosion/projects/tree/v3/tutorials/ner_double) that covers using multiple trained NER components for an example of how ordering matters.

> but I don't know how to match case-insensitive, without having to do something like this for each item in my list:

You can [specify the attribute](https://spacy.io/api/phrasematcher#:~:text=the%20token%20attribute%20to%20match%20on.) the PhraseMatcher should match on, in your case you can just use `LOWER`.
